src/Dialect/ONNX/ONNXOps/Canonicalize.cpp

```diff
@@ -1532,6 +1551,118 @@ struct PropagateBiasIntoLayerNormRewritePattern
   }
 };
 
+// =============================================================================
+// Rewrite pattern for Where
+// =============================================================================
+
+class RemoveWhereEqualPattern : public OpRewritePattern<ONNXWhereOp> {
+public:
+  using OpRewritePattern<ONNXWhereOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      ONNXWhereOp onnxWhereOp, PatternRewriter &rewriter) const override {
+    Location loc = onnxWhereOp.getLoc();
+    // Check operation pattern:
+    // (ONNXWhereOp
+    //     (ONNXEqualOp (ONNXConcatOp), (ONNXConstantOp)),
+    //      (ONNXConstantOp),
+    //      (ONNXConcatOp))
+    // - The second input of EqualOp need to be all negative values.
+    // - The output need to be integer type.
+    // - Has shape and rank.
+    // - DefiningOp of operands of ONNXConcatOp need to be DimOp or ConstantOp
+    // with scalar tensor
+    // - Operands in ONNXConcatOp need to be DimOp or ConstantOp
+    Value cond = onnxWhereOp.getCondition();
+    Value X = onnxWhereOp.getX();
+    Value Y = onnxWhereOp.getY();
+    if (mlir::isa<BlockArgument>(cond) || mlir::isa<BlockArgument>(X) ||
+        mlir::isa<BlockArgument>(Y))
+      return failure();
+
+    ONNXEqualOp equalOpCond = dyn_cast<ONNXEqualOp>(cond.getDefiningOp());
```

Comment:
Better to use `ONNXEqualOp equalOpCond = cond.getDefiningOp<ONNXEqualOp>();`

Reply:
There are also functions that does many of the steps here.
```
template <typename OP>
bool operandOfOpDefinedBy(mlir::Operation *&matchOp, mlir::Operation *op,
    mlir::Value &matchOperand, int64_t matchThisOperandIndex = 0);
```

for a given `op`, test that operand `matchThisOperandIndex` is defined by an op of type `typename OP`. If that is so, set that defining op into `matchOp` and the operand into `matchOperand` and return true.

It makes the code more readable. 

There is also a call that looks for one match out of 2 in a binary operation, if that is a pattern that you are interested in.

Reply:
Done. Thanks!